src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -2205,18 +2205,28 @@ internal Node RemoveAll(Predicate<T> match)
                 Contract.Ensures(Contract.Result<Node>() != null);
 
                 var result = this;
-                int index = 0;
-                foreach (var item in this)
+                var enumerator = new Enumerator(result);
+                try
                 {
-                    if (match(item))
+                    var startIndex = 0;
+                    while(enumerator.MoveNext())
                     {
-                        result = result.RemoveAt(index);
-                    }
-                    else
-                    {
-                        index++;
+                        if (match(enumerator.Current))
+                        {
+                            result = result.RemoveAt(startIndex);
+                            enumerator.Dispose();
```

Comment:
I have confirmed that `ImmutableList<T>.Enumerator.Dispose()` is safe for multiple calls.